Add XCTS system and tests

nilsvu opened the pull request:
## Proposed changes

Add the XCTS system and verifies the Schwarzschild solution solves it numerically.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu committed:
Add XCTS system

nilsvu committed:
Add test helper to verify XCTS solutions

This is to numerically verify the solution solves the system.

nilsvu committed:
Verify XCTS solutions solve system numerically

nilsvu commented:
I could move all commits except the first to a separate PR, but I kept them here for now because they test the XCTS system is implemented correctly.

nilsvu commented:
@haraldp271 and @nilsdeppe Would you be willing to continue the XCTS system review in this PR?

nilsdeppe commented:
sure, but I'm going to wait until the dependent PRs are merged since it's unclear which commits I should be reviewing

nilsdeppe commented on src/Elliptic/Systems/Xcts/FluxesAndSources.cpp:
```diff
@@ -0,0 +1,988 @@
+          const Scalar<DataVector>& /*extrinsic_curvature_trace*/,
+          const Scalar<DataVector>&
+          /*longitudinal_shift_minus_dt_conformal_metric_over_lapse_square*/,
+          const Scalar<DataVector>& /*conformal_factor*/) noexcept {}
```
[optional] in these empty functions, maybe add a brief comment here why they're empty? Just so that when someone looks at the source file there's a quick reminder

nilsdeppe commented on src/Elliptic/Systems/Xcts/FirstOrderSystem.hpp:
```diff
@@ -0,0 +1,261 @@
+ *
+ * \f{align}
+ * F^i_\psi &= \bar{\gamma}^{ij} v_j \\
+ * S_\psi &= -\bar{Gamma}^i_{ij}\bar{\gamma}^{jk}v_k
```
missing `\` in front of `Gamma`

nilsdeppe commented on src/Elliptic/Systems/Xcts/FirstOrderSystem.hpp:
```diff
@@ -0,0 +1,261 @@
+ * - \frac{1}{8}\psi^{-7}\bar{A}^2 - 2\pi\psi^5\rho \\
+ * f_\psi &= 0 \\
+ * F^i_{\beta^j} &= \left(\bar{L}\beta\right)^{ij} \\
+ * S_{\beta^i} &= -\bar{\Gamma}^j_{jk} \left(\bar{L}\beta\right)^{ik}
```
considering breaking this up over two lines

nilsvu replied:
👍 reformatted these equations

nilsdeppe commented on src/Elliptic/Systems/Xcts/FirstOrderSystem.hpp:
```diff
@@ -0,0 +1,261 @@
+ * + 16\pi\left(\alpha\psi\right)\psi^3 S^i \\
+ * f_{\beta^i} &= 0 \\
+ * F^i_{\alpha\psi} &= \bar{\gamma}^{ij} w^j \\
+ * S_{\alpha\psi} &= -\bar{\Gamma}^i_{ij}\bar{\gamma}^{jk}w_k
```
considering breaking this up over two lines

nilsdeppe commented on src/Elliptic/Systems/Xcts/FirstOrderSystem.hpp:
```diff
@@ -0,0 +1,261 @@
+ * \left(\bar{L}\beta\right)^{ij} +
+ * \left(\bar{L}\beta_\mathrm{background}\right)^{ij} - \bar{u}^{ij} \right) \\
+ * \text{and} \quad \left(\bar{L}\beta\right)^{ij} &=
+ * 2\left(\bar{gamma}^{ik}\bar{gamma}^{jl} - \frac{1}{3}
```
4 missing `\` in front of `gamma`s

nilsdeppe commented on src/Elliptic/Systems/Xcts/FirstOrderSystem.hpp:
```diff
@@ -0,0 +1,261 @@
+ * source terms as \f$\bar{\rho}=\psi^n\rho\f$, \f$\bar{S}=\psi^n S\f$ and
+ * \f$\bar{S^i}=\psi^n S^i\f$ and treat the re-scaled fields as
+ * freely-specifyable background data for the XCTS equations. You can select
+ * the `ConformalMatterScale` \f$n\f$ as the third template parameter.
```
question, is there a "Right" value of `n` or does that depend on either the physical setup or the matter type?

nilsvu replied:
I added a note to the docs. https://arxiv.org/abs/0804.3787 suggests n=6 to make volume integrals independent of the conformal factor (see footnote 1). @haraldp271 would you like to add some insight? :)

haraldp271 replied:
The big thing is to have n big enough for the correct sign.  Beyond that, I am not aware of a strongly preferred value (n=6 is nice, as the  integrated matter density independent of conformal factor)

nilsdeppe reviewed (commented):
LGTM, much easier to review than expected! Some dox+comment suggestions. Please feel free to squash immediately :)

nilsvu committed:
Fix a typo in longitudinal operator docs

nilsvu commented:
Rebased and squashed in the changes to the docs. Also added the small commit "Fix a typo in longitudinal operator docs". Thanks for reviewing @nilsdeppe !

nilsvu commented:
Rebased and added a fixup commit that verifies the freshly merged Kerr solution (#3068) :)

nilsdeppe reviewed (dismissed):
LGTM

nilsvu commented:
Rebased again. @haraldp271 Could you review this when you find the time?

haraldp271 reviewed (approved):
Looks good to me.  I keep on being surprised by the large amount of scaffolding codes that the templates necessitate.
